pkg/datastore/projectstore.go

```diff
@@ -201,15 +200,17 @@ func (s *projectStore) UpdateProjectRBACConfig(ctx context.Context, id string, r
 
 // UpdateProjectRBACRoles updates project rbac roles.
 func (s *projectStore) UpdateProjectRBACRoles(ctx context.Context, id string, roles []*model.ProjectRBACRole) error {
-	return ErrUnimplemented
+	return s.update(ctx, id, func(p *model.Project) error {
+		p.RbacRoles = roles
+		p.RbacRoles = p.FilterRBACRoles(false)
```

Comment:
tbh, I don't really like this approach. This leaks the fact that we only add `roles` which are not built-in roles, and from the caller side (which is not yet added in this PR), looks like we only have an interface like `projectstore.UpdateProjectRBACRoles` and there is no way to know what does it use for and will the passed roles are all applied or not. How about renaming this to `UpdateProjectRBACUserSettingRoles` and adding comments for it. Besides, I think roles are built-in or not should be encapsulated in project model (just like we made several private built-in roles variables in the project model package), so call to `p.FilterRBACRoles(isBuiltin=false)` from outside is a leak, should make `filterRBACRoles` as private and public only `p.Add/UpdateUserSettingsRoles()` and do filter internally. wdyt?

Reply:
> How about renaming this to UpdateProjectRBACUserSettingRoles and adding comments for it. 

Umm, if `UpdateProjectRBACRoles` has already existed, I think that makes sense but hasn't.
So I think it's okay that just add comments.

> p.FilterRBACRoles(isBuiltin=false) from outside is a leak, should make filterRBACRoles as private and public only p.Add/UpdateUserSettingsRoles() and do filter internally. wdyt?

Sorry, you've lost me.
Could you explain it in more detail?

Reply:
> p.FilterRBACRoles(isBuiltin=false) from outside is a leak, should make filterRBACRoles as private and public only p.Add/UpdateUserSettingsRoles() and do filter internally. 

I mean, since we have to know whether to pass isBuiltin value (true or false) from the caller side (datastore layer), it means that `is a given role a builtin or not` knowledge is leaked. So I don't think making this FilterRBACRoles interface a public interface is good. I guess, it's better to make this to a private method and public interface like p.UpdateUserSettingsRoles which accepts roles as its input and does map roles to p.RbacRoles and filterRBACRoles(false) as we did here.

Reply:
> There is a background of this implementation has the purpose of not being aware of whether it's built-in role or not from gateway anyway.

Knowledge leak means that we have to decide and control the inner logic by flag/params from the caller side, so I think to call to something like `p.FilterRBACRoles(isBuiltin=false)` is a knowledge leak issue. On the other hand, making an interface like `p.UpdateUserSettingsRoles` is not be counted as a knowledge leak, because the logic of handling whether a role is builtin or not and what we do to it is encapsulated inside the package via a public interface. That's my point 👀  Please correct me if I'm wrong.